File: Shared/Store/FavoritesStore.swift
import Foundation
import Observation
import CoreLocation

enum FavoriteRole: String, Codable, Hashable {
    case none, home, work
}

// Modo de transporte de un favorito.
enum FavoriteMode: String, Codable, Hashable {
    case tren, subte, bondi, bici
}

// Favorito de cualquier modo. Guarda todo lo necesario para mostrarlo y navegar,
// sin depender de catálogos (bondi y bici no tienen catálogo estático embebido).
struct FavoriteItem: Codable, Identifiable, Hashable {
    let mode: FavoriteMode
    let refId: String          // id estable por modo
    var name: String
    var lat: Double
    var lng: Double
    var lineLabel: String?     // texto del badge de línea (nil en bondi/bici)
    var lineColorHex: String?  // color del badge
    var routeId: String?       // subte: routeId para navegar
    var role: FavoriteRole
    var addedAt: Date

    var id: String { "\(mode.rawValue):\(refId)" }
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }
}

extension FavoriteItem {
    static func train(_ s: Station) -> FavoriteItem {
        FavoriteItem(mode: .tren, refId: String(s.id), name: s.nombre, lat: s.lat, lng: s.lng,
                     lineLabel: s.line.shortCode, lineColorHex: s.line.colorHex, routeId: nil,
                     role: .none, addedAt: Date())
    }
    static func subte(_ s: SubteStation) -> FavoriteItem {
        FavoriteItem(mode: .subte, refId: s.id, name: s.name, lat: s.lat, lng: s.lng,
                     lineLabel: s.line.letra, lineColorHex: s.line.colorHex, routeId: s.line.routeId,
                     role: .none, addedAt: Date())
    }
    static func bondi(_ r: ObaStopRef) -> FavoriteItem {
        FavoriteItem(mode: .bondi, refId: r.stopId, name: r.name, lat: r.lat, lng: r.lng,
                     lineLabel: nil, lineColorHex: nil, routeId: nil, role: .none, addedAt: Date())
    }
    static func bici(_ s: EcobiciStation) -> FavoriteItem {
        FavoriteItem(mode: .bici, refId: s.id, name: s.displayName, lat: s.lat, lng: s.lng,
                     lineLabel: nil, lineColorHex: nil, routeId: nil, role: .none, addedAt: Date())
    }
}

// Formato viejo (solo trenes) para migrar sin perder favoritos guardados.
private struct LegacyFavoriteStation: Codable {
    let stationId: Int
    var role: FavoriteRole
    var addedAt: Date
}

// Decodifica un elemento tolerando fallas: value == nil si ese elemento no decodifica.
private struct FailableDecodable<T: Decodable>: Decodable {
    let value: T?
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        value = try? container.decode(T.self)
    }
}

// Favoritos multi-modo con rol de contexto. Persiste en App Group.
@Observable
final class FavoritesStore {
    static let shared = FavoritesStore()

    private let defaults: UserDefaults
    private let key = "favorites.v2"
    private let legacyKey = "favorites.v1"
    private let catalog: StationCatalog

    private(set) var items: [FavoriteItem] = []

    init(defaults: UserDefaults? = nil, catalog: StationCatalog = .shared) {
        self.defaults = defaults ?? (UserDefaults(suiteName: "group.com.alandaitch.anden") ?? .standard)
        self.catalog = catalog
        load()
    }

    private func load() {
        // Si v2 existe es la fuente de verdad. Si el decode falla NO migramos v1
        // por encima (pisaría datos nuevos): dejamos vacío y no tocamos el disco.
        if let data = defaults.data(forKey: key) {
            if let decoded = try? JSONDecoder().decode([FavoriteItem].self, from: data) {
                items = decoded
            }
            return
        }
        migrateLegacy()
    }

    // Migra los favoritos v1 (solo trenes) al modelo nuevo y persiste en v2.
    // Solo corre cuando no existe v2. Decodifica elemento por elemento: un registro
    // corrupto no tira toda la migración.
    private func migrateLegacy() {
        guard let data = defaults.data(forKey: legacyKey) else { return }
        let legacy: [LegacyFavoriteStation]
        if let all = try? JSONDecoder().decode([LegacyFavoriteStation].self, from: data) {
            legacy = all
        } else if let raw = try? JSONDecoder().decode([FailableDecodable<LegacyFavoriteStation>].self, from: data) {
            legacy = raw.compactMap { $0.value }
        } else {
            legacy = []
        }
        if legacy.isEmpty {
            // v1 ilegible o vacío: limpiamos para no reintentar en cada arranque.
            defaults.removeObject(forKey: legacyKey)
            return
        }
        let migrated: [FavoriteItem] = legacy.compactMap { l in
            guard let st = catalog.station(id: l.stationId) else { return nil }
            return FavoriteItem(mode: .tren, refId: String(l.stationId),
                                name: st.nombre, lat: st.lat, lng: st.lng,
                                lineLabel: st.line.shortCode, lineColorHex: st.line.colorHex,
                                routeId: nil, role: l.role, addedAt: l.addedAt)
        }
        // Ninguno resolvió (¿catálogo no cargado?): no persistas ni borres, reintentá.
        if migrated.isEmpty { return }
        items = migrated
        persist()
        defaults.removeObject(forKey: legacyKey)
    }

    private func persist() {
        if let data = try? JSONEncoder().encode(items) {
            defaults.set(data, forKey: key)
        }
    }

    // MARK: - API por (mode, refId)

    func isFavorite(_ mode: FavoriteMode, _ refId: String) -> Bool {
        items.contains { $0.mode == mode && $0.refId == refId }
    }

    func toggle(_ item: FavoriteItem) {
        if let idx = items.firstIndex(where: { $0.mode == item.mode && $0.refId == item.refId }) {
            items.remove(at: idx)
        } else {
            items.append(item)
        }
        persist()
    }

    func remove(_ mode: FavoriteMode, _ refId: String) {
        items.removeAll { $0.mode == mode && $0.refId == refId }
        persist()
    }

    // Fija home/work sobre un favorito existente.
    func setRole(_ role: FavoriteRole, _ mode: FavoriteMode, _ refId: String) {
        guard let idx = items.firstIndex(where: { $0.mode == mode && $0.refId == refId }) else { return }
        items[idx].role = role
        persist()
    }

    func role(_ mode: FavoriteMode, _ refId: String) -> FavoriteRole {
        items.first { $0.mode == mode && $0.refId == refId }?.role ?? .none
    }

    // MARK: - Consumidores

    // Favorito "del momento": home a la mañana (4-12), work a la tarde (12-22).
    // Si no hay rol para la franja, el primero agregado. `among` limita los modos
    // (el widget pide solo modos con arribos: tren/subte/bondi).
    func contextualPrimary(now: Date = Date(), among modes: Set<FavoriteMode>? = nil) -> FavoriteItem? {
        let pool = modes.map { m in items.filter { m.contains($0.mode) } } ?? items
        let hour = Calendar.current.component(.hour, from: now)
        let preferred: FavoriteRole = (hour >= 4 && hour < 12) ? .home : ((hour >= 12 && hour < 22) ? .work : .none)
        if preferred != .none, let fav = pool.first(where: { $0.role == preferred }) {
            return fav
        }
        return pool.first
    }

    // Estaciones de tren favoritas resueltas (para notificaciones de demora, que hoy son solo tren).
    var trainStations: [Station] {
        items.filter { $0.mode == .tren }.compactMap { catalog.station(id: Int($0.refId) ?? -1) }
    }
}

File: android/app/src/main/java/com/alandaitch/anden/data/store/FavoritesStore.kt
package com.alandaitch.anden.data.store

import android.content.Context
import android.content.SharedPreferences
import com.alandaitch.anden.AndenApp
import com.alandaitch.anden.data.catalog.StationCatalog
import com.alandaitch.anden.data.model.EcobiciStation
import com.alandaitch.anden.data.catalog.SubteStation
import com.alandaitch.anden.data.model.Station
import com.alandaitch.anden.util.GeoPoint
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.Instant
import java.util.Calendar

enum class FavoriteRole { NONE, HOME, WORK }

// Modo de transporte de un favorito.
enum class FavoriteMode { TREN, SUBTE, BONDI, BICI }

// Favorito de cualquier modo. Guarda todo lo necesario para mostrarlo y navegar,
// sin depender de catálogos (bondi y bici no tienen catálogo estático embebido).
@Serializable
data class FavoriteItem(
    val mode: FavoriteMode,
    val refId: String,
    val name: String,
    val lat: Double,
    val lng: Double,
    val lineLabel: String? = null,
    val lineColorHex: String? = null,
    val routeId: String? = null,
    var role: FavoriteRole = FavoriteRole.NONE,
    val addedAtEpoch: Long = Instant.now().epochSecond,
) {
    val id: String get() = "${mode.name}:$refId"
    val coordinate: GeoPoint get() = GeoPoint(lat, lng)

    companion object {
        fun train(s: Station) = FavoriteItem(
            FavoriteMode.TREN, s.id.toString(), s.nombre, s.lat, s.lng,
            s.line.shortCode, s.line.colorHex,
        )
        fun subte(s: SubteStation) = FavoriteItem(
            FavoriteMode.SUBTE, s.id, s.name, s.lat, s.lng,
            s.line.letra, s.line.colorHex, s.line.routeId,
        )
        fun bondi(stopId: String, name: String, point: GeoPoint) = FavoriteItem(
            FavoriteMode.BONDI, stopId, name, point.lat, point.lng,
        )
        fun bici(s: EcobiciStation) = FavoriteItem(
            FavoriteMode.BICI, s.id, s.displayName, s.lat, s.lng,
        )
    }
}

// Formato viejo (solo trenes) para migrar sin perder favoritos guardados.
@Serializable
private data class LegacyFavoriteStation(
    val stationId: Int,
    val role: FavoriteRole = FavoriteRole.NONE,
    val addedAtEpoch: Long = 0,
)

// Favoritos multi-modo con rol de contexto. Persiste en SharedPreferences (JSON).
class FavoritesStore(
    context: Context = AndenApp.appContext,
    private val catalog: StationCatalog = StationCatalog.shared,
) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("anden.favorites", Context.MODE_PRIVATE)
    private val key = "favorites.v2"
    private val legacyKey = "favorites.v1"
    private val json = Json { ignoreUnknownKeys = true }

    private val _items = MutableStateFlow<List<FavoriteItem>>(emptyList())
    val itemsFlow: StateFlow<List<FavoriteItem>> = _items.asStateFlow()
    val items: List<FavoriteItem> get() = _items.value

    init {
        load()
    }

    private fun load() {
        prefs.getString(key, null)?.let { data ->
            runCatching { json.decodeFromString<List<FavoriteItem>>(data) }
                .getOrNull()?.let { _items.value = it; return }
        }
        migrateLegacy()
    }

    // Migra los favoritos v1 (solo trenes) al modelo nuevo y persiste en v2.
    private fun migrateLegacy() {
        val data = prefs.getString(legacyKey, null) ?: return
        val old = runCatching { json.decodeFromString<List<LegacyFavoriteStation>>(data) }.getOrNull() ?: return
        val migrated = old.mapNotNull { legacy ->
            val st = catalog.station(legacy.stationId) ?: return@mapNotNull null
            FavoriteItem(
                FavoriteMode.TREN, legacy.stationId.toString(), st.nombre, st.lat, st.lng,
                st.line.shortCode, st.line.colorHex, role = legacy.role, addedAtEpoch = legacy.addedAtEpoch,
            )
        }
        if (migrated.isNotEmpty()) {
            _items.value = migrated
            persist()
        }
    }

    private fun persist() {
        prefs.edit().putString(key, json.encodeToString(_items.value)).apply()
    }

    fun isFavorite(mode: FavoriteMode, refId: String): Boolean =
        _items.value.any { it.mode == mode && it.refId == refId }

    fun toggle(item: FavoriteItem) {
        val current = _items.value.toMutableList()
        val idx = current.indexOfFirst { it.mode == item.mode && it.refId == item.refId }
        if (idx >= 0) current.removeAt(idx) else current.add(item)
        _items.value = current
        persist()
    }

    fun remove(mode: FavoriteMode, refId: String) {
        _items.value = _items.value.filterNot { it.mode == mode && it.refId == refId }
        persist()
    }

    fun setRole(role: FavoriteRole, mode: FavoriteMode, refId: String) {
        val current = _items.value.toMutableList()
        val idx = current.indexOfFirst { it.mode == mode && it.refId == refId }
        if (idx < 0) return
        current[idx] = current[idx].copy(role = role)
        _items.value = current
        persist()
    }

    fun role(mode: FavoriteMode, refId: String): FavoriteRole =
        _items.value.firstOrNull { it.mode == mode && it.refId == refId }?.role ?: FavoriteRole.NONE

    // Favorito "del momento": home a la mañana (4-12), work a la tarde (12-22).
    // Si no hay rol para la franja, el primero agregado.
    fun contextualPrimary(now: Instant = Instant.now()): FavoriteItem? {
        val cal = Calendar.getInstance().apply { timeInMillis = now.toEpochMilli() }
        val hour = cal.get(Calendar.HOUR_OF_DAY)
        val preferred = when {
            hour in 4..11 -> FavoriteRole.HOME
            hour in 12..21 -> FavoriteRole.WORK
            else -> FavoriteRole.NONE
        }
        if (preferred != FavoriteRole.NONE) {
            _items.value.firstOrNull { it.role == preferred }?.let { return it }
        }
        return _items.value.firstOrNull()
    }

    companion object {
        val shared: FavoritesStore by lazy { FavoritesStore() }
    }
}
